### PR TITLE
Create set-based interface for NodesInfoRequest

### PR DESCRIPTION
**Pull request by williamrandolph**
This is another refactoring PR to move the Nodes Information endpoint in the direction of pluggability (see #52975). In #53140, we changed the `NodesInfoRequest` class so that it serializes as a set of strings rather than a fixed list of booleans. In this PR, we update the interface of `NodesInfoRequest` so that it uses getters and setters that are more appropriate to this internal structure. For example, the old way of adding "os" metrics to a request would be to call `request.os(true)`. The new way of doing this is to call `request.addMetric("os")`. For the time being, the `NodesInfoRequest` has a `Metric` enum that serves as the source of truth for allowed metrics. I plan to replace this eventually with a pluggable solution.

The follow-up step for this PR is to update the `NodesInfoRequestBuilder` to have a matching set-based interface, but I am putting that off to another PR in order to keep the diff small and to avoid confusion between the two classes.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java [hunk 1]
```diff
@@ -84,174 +85,68 @@ public NodesInfoRequest clear() {
...
-     * @param includeMetric Whether or not to include a metric.
+     * Helper method for adding and removing metrics. Used when deserializing
+     * a NodesInfoRequest from an ordered list of booleans.
+     *
+     * @param addMetric Whether or not to include a metric.
      * @param metricName Name of the metric to include or remove.
      */
-    private void addOrRemoveMetric(boolean includeMetric, String metricName) {
-        if (includeMetric) {
+    private void optionallyAddMetric(boolean addMetric, String metricName) {
```
Comment (williamrandolph) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java]: This method is now only used for backwards-compatible deserialization, and doesn't need to be able to remove metrics anymore.

**Comment by williamrandolph** [comment]
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/default-distro

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java [hunk 2]
```diff
@@ -281,12 +176,12 @@ public void writeTo(StreamOutput out) throws IOException {
...
      * pluggable.
      */
-    enum Metrics {
+    public enum Metric {
         SETTINGS("settings"),
         OS("os"),
         PROCESS("process"),
         JVM("jvm"),
-        THREAD_POOL("threadPool"),
+        THREAD_POOL("thread_pool"),
```
Comment (williamrandolph) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java]: Ah, this change leads to a bwc problem. I'm going to make a separate PR to handle the bwc aspects of this, for simplicity.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java [hunk 3]
```diff
@@ -84,174 +85,68 @@ public NodesInfoRequest clear() {
...
-        return Metrics.PLUGINS.containedIn(requestedMetrics);
-    }
-
-    /**
-     * Should information about ingest be returned
-     * @param ingest true if you want info
-     */
-    public NodesInfoRequest ingest(boolean ingest) {
-        addOrRemoveMetric(ingest, Metrics.INGEST.metricName());
+    public NodesInfoRequest addMetrics(String... metrics) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java]: Why do we need 2 variants of this method?

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestNodesInfoAction.java [hunk 4]
```diff
@@ -108,16 +98,9 @@ static NodesInfoRequest prepareRequest(final RestRequest request) {
...
-            nodesInfoRequest.os(metrics.contains("os"));
-            nodesInfoRequest.process(metrics.contains("process"));
-            nodesInfoRequest.jvm(metrics.contains("jvm"));
-            nodesInfoRequest.threadPool(metrics.contains("thread_pool"));
-            nodesInfoRequest.transport(metrics.contains("transport"));
-            nodesInfoRequest.http(metrics.contains("http"));
-            nodesInfoRequest.plugins(metrics.contains("plugins"));
-            nodesInfoRequest.ingest(metrics.contains("ingest"));
-            nodesInfoRequest.indices(metrics.contains("indices"));
+            // disregard unknown metrics
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestNodesInfoAction.java]: Aren't unknown metrics an error case?

**Review by rjernst — COMMENTED** [review]
I have a couple questions

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java (hunk 3 above)
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java]: This one made some of the invocations "look cleaner" to me, but it's not at all necessary and I'm fine with removing it.

File: server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestNodesInfoAction.java (hunk 4 above)
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestNodesInfoAction.java]: The current behavior of the Nodes Info API is to silently disregard invalid metrics (or node ids). The Nodes Stats endpoint, on the other hand, will return an error if you request an unknown metric.

Arguably, the Nodes Info API _should_ return an error for unknown metrics, but that would go beyond just refactoring.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java (hunk 3 above)
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java]: I've removed it.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java [hunk 5]
```diff
@@ -84,174 +85,60 @@ public NodesInfoRequest clear() {
...
-    }
-
-    /**
-     * Should information about ingest be returned
-     * @param ingest true if you want info
+     * Add collection of metrics
      */
-    public NodesInfoRequest ingest(boolean ingest) {
-        addOrRemoveMetric(ingest, Metrics.INGEST.metricName());
+    public NodesInfoRequest addMetrics(Collection<String> metrics) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java]: I liked having the variatic version. Is there a reason we need a collection version? It would make it easier to read in the rest of the code if the Set building part was isolated to the internals of this method.

**Review by rjernst — COMMENTED** [review]
Looks good, just one more request.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java (hunk 5 above)
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java]: That makes sense. I'll swap to the variatic version.

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java [hunk 6]
```diff
@@ -84,174 +88,70 @@ public NodesInfoRequest clear() {
...
-     * @return true if information about plugins is requested
-     */
-    public boolean plugins() {
-        return Metrics.PLUGINS.containedIn(requestedMetrics);
-    }
-
-    /**
-     * Should information about ingest be returned
-     * @param ingest true if you want info
+     * Add collection of metrics
```
Comment (williamrandolph) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java]: TODO: Fix javadoc for accuracy

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java [hunk 7]
```diff
@@ -84,174 +87,70 @@ public NodesInfoRequest clear() {
...
-
-    /**
-     * Should information about ingest be returned
-     * @param ingest true if you want info
+     * Add multiple metrics
      */
-    public NodesInfoRequest ingest(boolean ingest) {
-        addOrRemoveMetric(ingest, Metrics.INGEST.metricName());
+    public NodesInfoRequest addMetrics(String... metrics) {
+        SortedSet<String> illegalMetrics = new TreeSet<>();
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java]: This can be simplified in the same way I suggested in the response PR?

**Review by rjernst — APPROVED** [review]
LGTM, one last request

File: server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java (hunk 7 above)
Comment (williamrandolph) [reply server/src/main/java/org/elasticsearch/action/admin/cluster/node/info/NodesInfoRequest.java]: Will do!

**Comment by williamrandolph** [comment]
Backported: https://github.com/elastic/elasticsearch/pull/54223
